### PR TITLE
feat: Support pooling of YieldInstruction for Coroutines #92

### PR DESCRIPTION
**Pull request by AlexWargon**
## Purpose of this PR
Added pool support of WaitForSeconds/WaitForSecondsRealTime/WaitUntil/WaitWhile YieldInstructions

## Testing status

No tests have been added.

### Manual testing status
Checked new API in Editor

### CODE REVIEW

**Review by PavelKlym — CHANGES_REQUESTED** [review]
I like overall API, but some impl are questionable. Let's discuss!

File: Runtime/Async/Coroutine/Pooled YieldInstruction/YieldPool.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            for (var i = 0; i < size; i++)
+            {
+                instructions[YieldType<T>.Index].Enqueue(new T());
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void BackToPool<T>(T pooled) where T : PooledYieldInstruction {
+            instructions[YieldType<T>.Index].Enqueue(pooled);
+        }
```
[the remark this answers is not in the document]
Comment (AlexWargon) [reply Runtime/Async/Coroutine/Pooled YieldInstruction/YieldPool.cs:30]: It's faster then typeof(T)

**Review by PavelKlym — CHANGES_REQUESTED** [review]
LGTM! Just let's format everything with Rider.

File: Runtime/Async/Coroutine/Pooled YieldInstruction/YieldPool.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
+using System.Runtime.CompilerServices;
+using System;
+using System.Collections.Generic;
+using StansAssets.Foundation.Patterns;
+
+namespace StansAssets.Foundation.Async
+{
+    public static class YieldPool 
+    {
+        private static Dictionary<Type, DefaultPool<PooledYieldInstruction>> m_instructions;
```
Comment (PavelKlym) [inline Runtime/Async/Coroutine/Pooled YieldInstruction/YieldPool.cs:10]:
```suggestion
        private static Dictionary<Type, ObjectPool<PooledYieldInstruction>> m_instructions;
```
Please replace DefaultPool with ObjectPool.

File: Runtime/Async/Coroutine/Pooled YieldInstruction/PooledYieldInstruction.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+using UnityEngine;
+
+namespace StansAssets.Foundation.Async
+{
+    public class PooledYieldInstruction : CustomYieldInstruction
```
Comment (PavelKlym) [inline Runtime/Async/Coroutine/Pooled YieldInstruction/PooledYieldInstruction.cs:5]:
```suggestion
    internal abstract class PooledYieldInstruction : CustomYieldInstruction 
```

**Review by PavelKlym — CHANGES_REQUESTED** [review]
Some feedback sent after review.

**Review by PavelKlym — APPROVED** [review]
LGTM! Great feature and nice implementation.

**Pull request merged**